src/main/scala/mesosphere/marathon/api/akkahttp/v2/GroupsController.scala

```diff
@@ -39,50 +63,101 @@ class GroupsController(electionService: ElectionService, infoService: GroupInfoS
 
   def appsList(groupId: PathId): Route = ???
 
-  def createGroup(groupId: PathId): Route = ???
+  def createGroup(groupId: PathId)(implicit identity: Identity): Route = {
+    (forceParameter & entity(as(groupUpdateUnmarshaller(groupId)))) { (force, groupUpdate) =>
+      val effectiveGroupId = groupUpdate.id.map(id => PathId(id).canonicalPath(groupId)).getOrElse(groupId)
+      val rootGroup = groupManager.rootGroup()
+
+      if (rootGroup.group(effectiveGroupId).isDefined) { // group already exists
+        reject(Rejections.ConflictingChange(Message(s"Group $effectiveGroupId is already created. Use PUT to change this group.")))
+      } else if (rootGroup.transitiveAppsById.get(effectiveGroupId).isDefined) { // app with the group id already exists
+        reject(Rejections.ConflictingChange(Message(s"An app with the path $effectiveGroupId already exists.")))
+      } else {
+        onComplete(updateOrCreate(groupId, groupUpdate, force)) {
+          case Success(deploymentResult) => complete((StatusCodes.OK, List(Headers.`Marathon-Deployment-Id`(deploymentResult.deploymentId)), deploymentResult))
+          case Failure(ex: IllegalArgumentException) => complete(StatusCodes.UnprocessableEntity -> ex.getMessage)
+          case Failure(ex) => throw ex
+        }
+      }
+    }
+  }
+
+  /**
+    * Until there is async version of updateRoot we have to block here
+    */
+  protected def result[T](fn: Awaitable[T]): T = Await.result(fn, config.zkTimeoutDuration)
+
+  @SuppressWarnings(Array("all")) // async/await
+  private def updateOrCreate(id: PathId, update: raml.GroupUpdate, force: Boolean)(implicit identity: Identity): Future[DeploymentResult] = async {
+    val version = Timestamp.now()
+
+    val effectivePath = update.id.map(PathId(_).canonicalPath(id)).getOrElse(id)
+    val deploymentPlan = await(groupManager.updateRoot(
+      id.parent, group => result(groupApiService.updateGroup(group, effectivePath, update, version)), version, force))
+    DeploymentResult(deploymentPlan.id, deploymentPlan.version.toOffsetDateTime)
+  }
 
   def updateGroup(groupId: PathId): Route = ???
 
   def deleteGroup(groupId: PathId): Route = ???
 
-  def listVersions(groupId: PathId): Route = ???
+  def listVersions(groupId: PathId)(implicit identity: Identity): Route = {
+    groupManager.group(groupId) match {
+      case Some(group) =>
+        authorized(ViewGroup, group).apply {
+          val versionsFuture = groupManager.versions(groupId).runWith(Sink.seq)
+          onSuccess(versionsFuture) { versions =>
+            complete(versions)
+          }
+        }
+      case None => reject(EntityNotFound.noGroup(groupId))
+    }
+  }
 
-  def versionDetail(groupId: PathId, version: String): Route = ???
+  def versionDetail(groupId: PathId, version: Timestamp)(implicit identity: Identity): Route = extractEmbeds {
+    case (appEmbed, groupEmbed) =>
```

Comment:
Why `case` here?

Reply:
because the input is tuple and I want to unwrap that...

Reply:
Ok, it was just a little confusing.